### PR TITLE
Pay with bat skus

### PR DESCRIPTION
**Pull request by husobee**
### Type of change ( select one )

- [ ] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: libs/clients/radom/client.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		if subtle.ConstantTimeCompare([]byte(key), []byte(v.Key)) == 1 {
+			return v.Value, nil
+		}
+	}
+	return nil, errors.New("failed to get key from radom metadata")
+}
+
+// WebhookRequest is the request from radom webhooks
+type WebhookRequest struct {
+	EventName            string   `json:"eventName"`
```
Comment (mariel222) [inline libs/clients/radom/client.go:33]: The webhook payload structure has changed with the managed payments API. There are 2 webhook messages you can listen on - either for a new subscription or a recurring payment. The recurring payment event would happen every time a charge happens on the subscription period basis - and the new subscription event would only trigger once (and a respective subscription cancellation event would trigger in the event it's cancelled).

```json
{
  "eventType": "managedRecurringPayment",
  "eventData": {
    "managedRecurringPayment": {
      "paymentMethod": {
        "network": "SepoliaTestnet",
        "token": "0x5d684d37922daf7aa2013e65a22880a11c475e25"
      },
      "amount": "10.149736300000000",
      "alternativeMethodPayments": null,
      "transactions": null
    }
  },
  "radomData": {
    "checkoutSession": {
      "checkoutSessionId": "17f3a621-ce07-4d38-883b-3d31622a4c9d",
      "metadata": [
        {
          "key": "brave-metadata",
          "value": "some-id"
        }
      ]
    }
  }
}
```

```json
{
  "eventType": "newSubscription",
  "eventData": {
    "newSubscription": {
      "subscriptionId": "945954fa-badc-498b-9978-7f1712c05cc1",
      "subscriptionType": {
        "automatedEVMSubscription": {
          "buyerAddress": "0x850f184effee1f86472c056b777126bd063e4cd0",
          "subscriptionContractAddress": "0x87Fcf739aEe21FAb3b99121fA4f2D7bF243f9F15",
          "recommendedAllowanceDuration": null
        }
      },
      "network": "SepoliaTestnet",
      "token": "0x5D684d37922dAf7Aa2013E65A22880a11C475e25",
      "amount": "1.9900",
      "currency": "USD",
      "period": "annually",
      "periodCustomDuration": null,
      "createdAt": "2023-07-12T21:20:13.458273Z",
      "tags": {
        "productId": "c33369a9-5afc-4818-9ab3-1793297f1e0e"
      }
    }
  },
  "radomData": {
    "checkoutSession": {
      "checkoutSessionId": "17f3a621-ce07-4d38-883b-3d31622a4c9d",
      "metadata": [
        {
          "key": "brave-metadata",
          "value": "some-id"
        }
      ]
    }
  }
}
```

Apologies for the big changes in the structure. Here is a link to our most recent docs in regards to these webhook messages: https://docs.radom.network/guides/webhooks

File: services/skus/model/model.go [hunk 2]
```diff
@@ -69,6 +79,65 @@ func (o *Order) IsStripePayable() bool {
...
+		CancelURL:      cancelURI,
+		Currency:       "BAT",
+		AcceptedTokens: acceptedTokens,
+		AcceptedChains: acceptedChains,
+		SellerAddress:  sellerAddress,
+		Metadata: radom.Metadata(
+			[]radom.KeyValue{
+				{
+					Key: "brave-metadata",
+					Value: map[string]interface{}{
```
Comment (mariel222) [inline services/skus/model/model.go:120]: This metadata value should be a string. I guess you could either serialize this map as a string if you're wanting to add more fields - or include them as separate objects in the array.

**Review by mariel222 — CHANGES_REQUESTED** [review]
LGTM overall. Just some comments from the API changes to the new managed payments API.

**Review by pavelbrm — COMMENTED** [review]
Added a couple of comments on the sku files, plus expressed my feedback in #1892.

File: tools/macaroon/cmd/brave-firewall-vpn/premium_dev_time_limited_v2_bat.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+tokens:
+  - id: "brave-vpn-premium"
+    version: 1
+    location: "vpn.brave.software"
+    first_party_caveats:
+      - sku: "brave-firewall-vpn-premium"
```
[the remark this answers is not in the document]
Comment (husobee) [reply tools/macaroon/cmd/brave-firewall-vpn/premium_dev_time_limited_v2_bat.yaml:6]: legacy naming issue which can be fixed with bundling and non macaroon skus

File: tools/macaroon/cmd/brave-firewall-vpn/premium_dev_time_limited_v2_bat.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      - price: 15
+      - currency: "BAT"
+      - description: "Brave Firewall + VPN"
```
[the remark this answers is not in the document]
Comment (husobee) [reply tools/macaroon/cmd/brave-firewall-vpn/premium_dev_time_limited_v2_bat.yaml:9]: legacy naming issue which can be fixed with bundling and non macaroon skus

File: libs/clients/radom/radom.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	"errors"
+	"time"
+
+	"github.com/shopspring/decimal"
+
+	"github.com/brave-intl/bat-go/libs/clients"
+)
+
+var (
+	ErrInvalidMetadataKey = errors.New("failed to get key from radom metadata")
```
Comment (clD11) [inline libs/clients/radom/radom.go:15]: nit: maybe message could read `invalid metadata key`

**Review by clD11 — APPROVED** [review]

File: services/skus/model/model.go [hunk 6]
```diff
@@ -126,6 +143,98 @@ func (o *Order) CreateStripeCheckoutSession(
...
+	} else {
+		// append production
+		gateway.Managed.Methods = append(gateway.Managed.Methods, radom.Method{
+			Network: "Polygon",
+			Token:   "0x3cef98bb43d732e2f285ee605a8158cde967d219",
+		}, radom.Method{
+			Network: "Ethereum",
+			Token:   "0x0d8775f648430679a709e98d2b0cb6250d2887ef",
+		})
+	}
```
Comment (pavelbrm) [inline services/skus/model/model.go:188]: Non-blocking: Ideally, these should take place after we've done the validation steps below, but before making a request. I.e. there is little point in initialising this if the order itself is not worth action on.

**Review by pavelbrm — APPROVED** [review]
Looking good. 👍

**Comment by husobee** [comment]
https://github.com/brave/reviews/issues/1357

File: libs/clients/radom/radom.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+var (
+	ErrInvalidMetadataKey = errors.New("invalid metadata key")
+)
+
+// CheckoutSessionRequest represents a request to create a checkout session.
+type CheckoutSessionRequest struct {
+	SuccessURL     string                 `json:"successUrl"`
+	CancelURL      string                 `json:"cancelUrl"`
+	Currency       string                 `json:"currency"`
+	ExpiresAt      int64                  `json:"expiresAt"` // in unix seconds
```
Comment (fmarier) [inline libs/clients/radom/radom.go:23]: What do we set this to when we create a Radom session?

Comment (husobee) [reply libs/clients/radom/radom.go:23]: these are set from the sku values, and the success/cancel urls are the brace account site, currency is BAT and expires at is the end of the sku duration.  here is an example of what we set:

https://github.com/brave-intl/bat-go/pull/1789/files#diff-30be4362fa150d72b34eede7fddf4e7bc71aa79064ae0dc514bc311b613ab19aR11-R19

File: services/skus/controllers.go [hunk 8]
```diff
@@ -1041,6 +1046,91 @@ func HandleIOSWebhook(service *Service) handlers.AppHandler {
...
+			return handlers.WrapError(err, "invalid verification key", http.StatusBadRequest)
+		}
+
+		req := radom.WebhookRequest{}
+		if err := requestutils.ReadJSON(ctx, r.Body, &req); err != nil {
+			lg.Error().Err(err).Msg("failed to read request body")
+			return handlers.WrapError(err, "error reading request body", http.StatusServiceUnavailable)
+		}
+
+		lg.Debug().Str("event_type", req.EventType).Str("data", fmt.Sprintf("%+v", req)).Msg("webhook event captured")
```
[the remark this answers is not in the document]
Comment (husobee) [reply services/skus/controllers.go:1075]: no, there is no personal data in this data blob from radom, and in production we have log level set to info not debug.

**Review by fmarier — APPROVED** [review]

**Comment by husobee** [comment]
> Please do not merge. Somehow, the recent conflict resolution brought in back code that has been removed. We need to fix that.
> 
> cc @husobee

@pavelbrm I have corrected the merge faults.  Please take a look when you can?

File: services/skus/model/model.go (hunk 6 above)
Comment (pavelbrm) [reply services/skus/model/model.go]: I moved gateway management inside client. Upon a client method call (create session), the client will determine the appropriate gateway settings and attach them to the request.

This knowledge about the gateway does not belong in the order context. It is also not different from order to order. That means it does not have to be here, and it has to be an internal to the client detail.

Plus, allocating slices for every single order call is not necessary, and we can reuse the same gateway methods again and again.

Pushed that change.

**Comment by pavelbrm** [comment]
> > Please do not merge. Somehow, the recent conflict resolution brought in back code that has been removed. We need to fix that.
> > cc @husobee
> 
> @pavelbrm I have corrected the merge faults. Please take a look when you can?

Thanks @husobee ! I have deleted the code that had to be deleted (some of the fixes have brought in the old implementation of payment methods). Also, addressed some of my comments on gateway management (moved that to the client).

As soon as I see tests pass, will re-approve the PR.

**Pull request merged**